### PR TITLE
CM - 97 Swagger error messages

### PR DESCRIPTION
**Pull request by Farhad-Rezwan**
Screen Shot 2022-04-01 at 4 52 02 pm

### CODE REVIEW

File: app/Controllers/Http/AssetsController.ts [hunk 1]
```diff
@@ -49,28 +52,36 @@ export default class AssetsController {
...
+    if (!user || user.companyId !== jwtAccount.companyId)
+      throw new UnauthorizedAccessException('Unauthorized access', StatusCodes.UnauthorizedAccess);
 
     const company = await Company.query().where('id', user.companyId).preload('ledgers').firstOrFail();
 
     const hasLedger = company.ledgers.some(({ id }) => asset.ledger.id === id);
 
     if (!hasLedger) {
-      throw new ForbiddenException('No permission to access asset ledger');
+      throw new ForbiddenException('No permission to access asset ledger', StatusCodes.Forbidden);
```
Comment (smallbatch-apps) [inline app/Controllers/Http/AssetsController.ts:84]: As above it's important to note that this is the response that goes back to the user. You're essentially saying that they're forbidden from accessing this resource. Why? They should have permission for this resource, it's their resource. Internally something is invalid, but the user doesn't know that. Essentially this needs to be some kind of `5xx` error.

File: docs/swagger/assets.yml [hunk 2]
```diff
@@ -55,10 +55,40 @@
             schema:
               type: object
               $ref: '#/components/schemas/Asset'
+      401:
+        description: Invalid authentication
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/errors/UnauthorizedAccess'
```
Comment (smallbatch-apps) [inline docs/swagger/assets.yml:63]: It's interesting that this is able to access errors from an external file. If that's how it's done maybe we can do the same with schemas as well?

**Review by smallbatch-apps — CHANGES_REQUESTED** [review]

File: app/Controllers/Http/AssetsController.ts (hunk 1 above)
Comment (Farhad-Rezwan) [reply app/Controllers/Http/AssetsController.ts:84]: Makes sense

File: docs/swagger/assets.yml (hunk 2 above)
Comment (Farhad-Rezwan) [reply docs/swagger/assets.yml:63]: Yup.

File: app/Controllers/Http/AssetsController.ts [hunk 3]
```diff
@@ -49,28 +52,36 @@ export default class AssetsController {
...
+    if (!user || user.companyId !== jwtAccount.companyId)
+      throw new UnauthorizedAccessException('Unauthorized access', StatusCodes.UnauthorizedAccess);
+
     const company = await Company.query().where('id', user.companyId).preload('ledgers').firstOrFail();
 
     let parentLedger = await LedgerLucidRepository.findParentLedger(company);
 
     if (!parentLedger) {
-      throw new LedgerNotFoundException('No Parent Ledger for company');
+      throw new LedgerNotFoundException('No Parent Ledger for company', StatusCodes.NotFound);
```
[the remark this answers is not in the document]
Comment (Farhad-Rezwan) [reply app/Controllers/Http/AssetsController.ts:65]: Makes sense

File: app/Controllers/Http/AssetsController.ts [hunk 4]
```diff
@@ -49,28 +52,36 @@ export default class AssetsController {
 
     Sentry.addBreadcrumb({ message: 'Creating Asset', level });
 
-    const user = await User.findOrFail(jwtAccount.userId);
+    const user = await User.find(jwtAccount.userId);
+
+    if (!user || user.companyId !== jwtAccount.companyId)
+      throw new UnauthorizedAccessException('Unauthorized access', StatusCodes.UnauthorizedAccess);
```
[the remark this answers is not in the document]
Comment (Farhad-Rezwan) [reply app/Controllers/Http/AssetsController.ts:58]: done

File: docs/swagger/assets.yml [hunk 5]
```diff
@@ -109,69 +145,23 @@
...
+      403:
+        description: Forbidden asset access
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/errors/Forbidden'
+            example:
+              message: 'No permission to access asset ledger'
+      404:
+        description: Company not found for user
```
Comment (thanapat-civicledger) [inline docs/swagger/assets.yml:163]: This is a bit confusing for me, as the same error code will be returned if no asset is found for the given asset id.
I don't know how to better handle this case. Maybe use the description as "Asset not found | Company not found for user"?

File: app/Controllers/Http/AssetsController.ts [hunk 6]
```diff
@@ -20,7 +20,10 @@ export default class AssetsController {
   public async index({ request, jwtAccount }: HttpContextContract) {
     const { entryName, ledgerId, ...otherQueryStrings } = request.qs();
 
-    const user = await User.findOrFail(jwtAccount.userId);
+    const user = await User.find(jwtAccount.userId);
+
+    if (!user || user.companyId !== jwtAccount.companyId)
```
Comment (smallbatch-apps) [inline app/Controllers/Http/AssetsController.ts:25]: Sorry, I missed this one, can you wrap it too? This sort of multiple line conditional is very error prone. It should be in brackets.

Comment (Farhad-Rezwan) [reply app/Controllers/Http/AssetsController.ts:25]: Oops sorry I missed it to.

Comment (Farhad-Rezwan) [reply app/Controllers/Http/AssetsController.ts:25]: Just updated

File: docs/swagger/assets.yml (hunk 5 above)
Comment (Farhad-Rezwan) [reply docs/swagger/assets.yml:163]: Yes, we can do that. This doc is just an example of one use case error, however I have updated accordingly.

**Review by thanapat-civicledger — APPROVED** [review]
Looks good to me

**Review by smallbatch-apps — APPROVED** [review]

**Pull request merged**